src/api.ts

```diff
@@ -0,0 +1,70 @@
+import { request } from "./common"
+
+export type SpawnRequestBody = {
+    env?: Record<string, string>; // env vars always map strings to strings
+    port?: number;
+    tag?: string;
+}
+
+// TODO(paul): formalize the API return types and create a public API spec (OpenAPI?)
+
+export class JamsocketApi {
+    apiBase: string
+
+    constructor(private auth: string) {
+        this.apiBase = process.env.JAMSOCKET_SERVER_API ?? 'https://jamsocket.dev'
+    }
+
+    private async makeAuthenticatedRequest(endpoint: string, body?: any): Promise<any> {
+        const url = `${this.apiBase}${endpoint}`;
+        const response = await request(url, body || null, {
+            method: body ? 'POST' : 'GET',
+            headers: { 'Authorization': `Basic ${this.auth}` },
+        });
+
+        let responseBody
+        try {
+            responseBody = JSON.parse(response.body);
+        } catch (error) {
+            throw new Error(`jamsocket: error parsing JSON response - ${error}: ${response.body}`);
+        }
+
+        if (response.statusCode && response.statusCode >= 400) {
+            const { message, status, code, id } = responseBody.error
+            throw new Error(`jamsocket: ${status} - ${code}: ${message} (id: ${id})`);
+        }
+
+        return responseBody
+    }
+
+    public async checkAuth(): Promise<void> {
+        const url = `/api/auth`;
+        await this.makeAuthenticatedRequest(url);
+    }
+
+    public async serviceImage(username: string, serviceName: string): Promise<string> {
+        const url = `/api/user/${username}/service/${serviceName}/image`;
+        const result = await this.makeAuthenticatedRequest(url);
+        return result['imageName']
```

Comment:
Not a big deal, but I noticed that with `serviceImage()` we're pulling out `imageName` from the response body and returning that while in other functions here we're passing the response body through expecting the caller to know the shape of the response body.

If we just passed the entire response body through here like we are elsewhere, we could remove some lines from each of these functions. We could actually just return `this.makeAuthenticatedRequest(url)` without awaiting it since we have to `await serviceImage()` anyway, the extra `await` here just before the end of the function doesn't actually do anything for us.

Awaiting before returning is the equivalent of doing this:

```js
this.makeAuthenticatedRequest.then(response => Promise.resolve(response))
```